draft-ietf-jose-hpke-encrypt.md

```diff
@@ -131,70 +131,172 @@ This specification uses the following abbreviations and terms:
+- pt: The message plaintext, compressed using the algorithm in "zip" header parameter if present.
+
+Additionally, if the "psk_id" header parameter is present:
+
+- psk: The pre-shared key.
+- psk_id: The value of the "psk_id" header parameter.
+
+Additionally, if the "auth_kid" header parameter is present:
+
+- skS: The sender private key, converted into HPKE private key.
+
+The resulting JWE is filled as follows:
+
+- JWE Protected Header:
+  * SHALL contain "enc" with value "dir"
+  * SHALL contain "alg" that is the used JOSE-HPKE algorithm.
+  * MUST contain the "apu", "apv" and "zip" header parameters, if present.
+- JWE Initialization Vector SHALL be empty.
+- JWE Ciphertext SHALL be the raw ct output from HPKE Seal operation.
+- JWE Authentication Tag SHALL be empty.
+- There MUST be exactly one recipient, with:
+  * JWE Per-Recipient Unprotected Header MUST be empty.
+  * JWE Encrypted Key SHALL contain the raw enc output from HPKE Seal operation.
+
+The "ek" header parameter MUST NOT be present.
+
+When decrypting, the inputs to HPKE Open operation are set as follows:
+
+- kem_id: Depends on the JOSE-HPKE algorithm used.
+- skR: The recipient private key, converted into HPKE private key.
+- kdf_id: Depends on the JOSE-HPKE algorithm used.
+- aead_id: Depends on the JOSE-HPKE algorithm used.
+- info: By default, an empty string. Application MAY specify some other value.
+- add: Additional Authenticated Data encryption parameter defined in {{RFC7516}} section 5.2. step 15.
+- enc: The JWE Encrypted Key of the sole recipient.
+- ct: The JWE Ciphertext.
+
+Additionally, if the "psk_id" header parameter is present:
+
+- psk: The pre-shared key.
+- psk_id: The value of the "psk_id" header parameter.
 
-JWE supports different serializations, including Compact JWE Serialization as described in Section 3.1 of {{RFC7516}}, General JWE JSON Serialization as described in Section 3.2 of {{RFC7516}}.
+Additionally, if the "auth_kid" header parameter is present:
 
-Certain JWE features are only supported in specific serializations.
+- pkS: The sender public key, converted into HPKE public key.
 
-For example Compact JWE Serialization does not support the following:
+If the "zip" header parameter is present, the resulting plaintext is uncompressed using the algorithm specified and the result is the
+raw message plaintext. Otherwise the resulting plaintext is the raw message plaintext.
 
-- additional authenticated data
-- multiple recipients
-- unprotected headers
+When decrypting, the checks in {{RFC7516}} section 5.2. steps 1 through 5 MUST be performed.
 
-HPKE JWE Key Encryption can be used with "aad" but only when not expressed with Compact JWE Serialization.
+## Key Encryption
 
-Single recipient HPKE JWE Key Encryption with no "aad" can be expressed in Compact JWE Serialization, so long as the recipient and sender use the same HPKE Setup process as described in { Section 5 of RFC9180 }.
+The Recipient Context is defined as follows:
```

Comment:
As aad input in the Key Encryption case.

Reply:
If apu and apv in JOSE are pretty much completely broken. why use it in the Recipient context ? 
What about the "JWE Additional Authenticated Data encryption parameter" defined in Step 14 of Section 5.1 of RFC7516 ?

Reply:
Apu and apv were included before, I guess to maintain parity with Key Agreement.

And both modes do include JWE AAD. It is part of "Additional Authenticated Data encryption parameter", which then gets fed as aad to bulk encryption (either HPKE or JWE native bulk encryption).

Reply:
The problem is JWE AAD is not included in the Key Encryption mode, it is only included in the Integrated Encryption mode.  It needs to be fixed; it was included in both modes in the latest published version of the draft. 

One possible solution for both modes would be use HPKE Setup info to carry the Recipient Context. In both modes, JWE AAD should be conveyed in HPKE aad.

Reply:
It **is** included in Key Encryption mode, as JWE implicitly includes it.

Integrated Encryption has to explicitly include it, because it bypasses the parts of JWE that normally ensure that JWE AAD gets included.

Reply:
In the revised Key Encryption Section, HPKE aad only carries the Recipient Context. JWE AAD needs to be conveyed in the HPKE aad.

Reply:
No, it does not.

Reply:
Yes, but why did you delete most of the existing text in Section 6 (Key Encryption) ?

Reply:
Most of that section was just repeating stuff from JWE, often not quite correctly.

Reply:
I don’t see any issues with removing the redundant content from JWE. However, I will need to review all the changes to ensure that no relevant text is removed. I noticed the following:

1. Section 4.2 is removed.
2. In Key Encryption section, the latest version of the draft says encrypted_key is base64url encoded content
    encryption key but in the PR it says encrypted_key carries the raw enc output.

Reply:
Regarding 1., that section was redundant with specified HPKE I/O mappings.

Regarding 2., Note that the PR says "JWE Encrypted Key", not "encrypted_key".  JWE specifies that the latter is base64url encoding of the former. So the field ends up base64url-encoded.

For the "ek" parameter, there is no automatic base64url-encoding, so the PR explicitly specifies base64url endec.